*****Finding the best predictors of covid death rates*****

Covid 19 has had a major impact on the world and it is important that we come to understand how we might be able to survive this disease. This project seeks to find the major risk factors that could lead to death from Covid 19. In this notebook, we look to just collect and clean the covid dataset. The country of focus is the United States.

**The first step is to import the necessary libraries.**

In [1]:
import pandas as pd

**Next is to bring the dataset in as a pandas dataframe using the method read.csv. The first 5 rows are then shown**

In [2]:
df = pd.read_csv('covid_data.csv', sep=';')

In [3]:
df.head()

iso_code continent     location        date  total_cases  new_cases  \
0      AFG      Asia  Afghanistan  2020/02/24          5.0        5.0   
1      AFG      Asia  Afghanistan  2020/02/25          5.0        0.0   
2      AFG      Asia  Afghanistan  2020/02/26          5.0        0.0   
3      AFG      Asia  Afghanistan  2020/02/27          5.0        0.0   
4      AFG      Asia  Afghanistan  2020/02/28          5.0        0.0   

   new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  ...  \
0                 NaN           NaN         NaN                  NaN  ...   
1                 NaN           NaN         NaN                  NaN  ...   
2                 NaN           NaN         NaN                  NaN  ...   
3                 NaN           NaN         NaN                  NaN  ...   
4                 NaN           NaN         NaN                  NaN  ...   

   female_smokers  male_smokers  handwashing_facilities  \
0             NaN           NaN                  37.746   
1             NaN           NaN                  37.746   
2             NaN           NaN                  37.746   
3             NaN           NaN                  37.746   
4             NaN           NaN                  37.746   

   hospital_beds_per_thousand  life_expectancy  human_development_index  \
0                         0.5            64.83                    0.511   
1                         0.5            64.83                    0.511   
2                         0.5            64.83                    0.511   
3                         0.5            64.83                    0.511   
4                         0.5            64.83                    0.511   

   excess_mortality_cumulative_absolute  excess_mortality_cumulative  \
0                                   NaN                          NaN   
1                                   NaN                          NaN   
2                                   NaN                          NaN   
3                                   NaN                          NaN   
4                                   NaN                          NaN   

   excess_mortality  excess_mortality_cumulative_per_million  
0               NaN                                      NaN  
1               NaN                                      NaN  
2               NaN                                      NaN  
3               NaN                                      NaN  
4               NaN                                      NaN  

[5 rows x 67 columns]

The Data has been successfully imported into the notebook. The data is from February 2020 to April 2022.

Looking to see the stats of the data

In [4]:
df.describe()

total_cases     new_cases  new_cases_smoothed  total_deaths  \
count  1.726410e+05  1.724630e+05        1.712940e+05  1.544800e+05   
mean   2.882052e+06  1.242727e+04        1.243800e+04  6.101233e+04   
std    1.791153e+07  8.872410e+04        8.708769e+04  3.190189e+05   
min    1.000000e+00  0.000000e+00        0.000000e+00  1.000000e+00   
25%    2.329000e+03  1.000000e+00        6.857000e+00  8.600000e+01   
50%    3.052600e+04  7.800000e+01        1.064290e+02  8.370000e+02   
75%    3.364600e+05  1.073000e+03        1.172571e+03  7.891250e+03   
max    5.036050e+08  4.089026e+06        3.436980e+06  6.195647e+06   

          new_deaths  new_deaths_smoothed  total_cases_per_million  \
count  154500.000000        153353.000000            171842.000000   
mean      166.240369           167.166598             34887.610170   
std       814.611919           797.676071             64162.041734   
min         0.000000             0.000000                 0.001000   
25%         0.000000             0.143000               684.185250   
50%         2.000000             2.286000              5391.156000   
75%        19.000000            20.857000             43577.160000   
max     18144.000000         14783.286000            706541.904000   

       new_cases_per_million  new_cases_smoothed_per_million  \
count          171664.000000                   170500.000000   
mean              181.871695                      181.769950   
std               749.035427                      577.338034   
min                 0.000000                        0.000000   
25%                 0.022750                        1.592000   
50%                11.164000                       19.233000   
75%               103.632750                      126.342000   
max             51427.491000                    16052.608000   

       total_deaths_per_million  ...  female_smokers   male_smokers  \
count             153694.000000  ...   112127.000000  110590.000000   
mean                 546.383599  ...       10.649534      32.786541   
std                  827.897023  ...       10.587423      13.523888   
min                    0.000000  ...        0.100000       7.700000   
25%                   20.384000  ...        1.900000      21.600000   
50%                  146.061500  ...        6.300000      31.400000   
75%                  784.131750  ...       19.300000      41.300000   
max                 6371.964000  ...       44.000000      78.100000   

       handwashing_facilities  hospital_beds_per_thousand  life_expectancy  \
count            72641.000000               130906.000000    167562.000000   
mean                50.843202                    3.027735        73.663241   
std                 31.809860                    2.447097         7.463400   
min                  1.188000                    0.100000        53.280000   
25%                 20.859000                    1.300000        69.500000   
50%                 49.839000                    2.400000        75.090000   
75%                 82.502000                    4.000000        79.190000   
max                100.000000                   13.800000        86.750000   

       human_development_index  excess_mortality_cumulative_absolute  \
count            144191.000000                          6.078000e+03   
mean                  0.725414                          3.899161e+04   
std                   0.149847                          1.088362e+05   
min                   0.394000                         -3.772610e+04   
25%                   0.602000                         -4.415000e+01   
50%                   0.743000                          3.585550e+03   
75%                   0.845000                          2.639225e+04   
max                   0.957000                          1.163660e+06   

       excess_mortality_cumulative  excess_mortality  \
count                  6078.000000       6078.000000   
mean                      9.582009         15.804225   


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 179218 entries, 0 to 179217
Data columns (total 67 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   iso_code                                    179218 non-null  object 
 1   continent                                   168729 non-null  object 
 2   location                                    179218 non-null  object 
 3   date                                        179218 non-null  object 
 4   total_cases                                 172641 non-null  float64
 5   new_cases                                   172463 non-null  float64
 6   new_cases_smoothed                          171294 non-null  float64
 7   total_deaths                                154480 non-null  float64
 8   new_deaths                                  154500 non-null  float64
 9   new_deaths_smoothed                         153353 non-null  float64
 

As can be seen above, this is a pretty large dataset. But it can also be noted that the data includes entries from other countries outside United States. This is out of the scope of this study. We are only looking for data from the United States.

Next, we subset the data to only show United States data as USdf

In [6]:
df.location.unique()

array(['Afghanistan', 'Africa', 'Albania', 'Algeria', 'Andorra', 'Angola',
       'Anguilla', 'Antigua and Barbuda', 'Argentina', 'Armenia', 'Aruba',
       'Asia', 'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain',
       'Bangladesh', 'Barbados', 'Belarus', 'Belgium', 'Belize', 'Benin',
       'Bermuda', 'Bhutan', 'Bolivia', 'Bonaire Sint Eustatius and Saba',
       'Bosnia and Herzegovina', 'Botswana', 'Brazil',
       'British Virgin Islands', 'Brunei', 'Bulgaria', 'Burkina Faso',
       'Burundi', 'Cambodia', 'Cameroon', 'Canada', 'Cape Verde',
       'Cayman Islands', 'Central African Republic', 'Chad', 'Chile',
       'China', 'Colombia', 'Comoros', 'Congo', 'Cook Islands',
       'Costa Rica', "Cote d'Ivoire", 'Croatia', 'Cuba', 'Curacao',
       'Cyprus', 'Czechia', 'Democratic Republic of Congo', 'Denmark',
       'Djibouti', 'Dominica', 'Dominican Republic', 'Ecuador', 'Egypt',
       'El Salvador', 'Equatorial Guinea', 'Eritrea', 'Estonia',
       'Eswatini', 'Ethi

In [7]:
USdf = df[df.location == 'United States']

In [8]:
USdf.head()

iso_code      continent       location        date  total_cases  \
168832      USA  North America  United States  2020/01/22          1.0   
168833      USA  North America  United States  2020/01/23          1.0   
168834      USA  North America  United States  2020/01/24          2.0   
168835      USA  North America  United States  2020/01/25          2.0   
168836      USA  North America  United States  2020/01/26          5.0   

        new_cases  new_cases_smoothed  total_deaths  new_deaths  \
168832        NaN                 NaN           NaN         NaN   
168833        0.0                 NaN           NaN         NaN   
168834        1.0                 NaN           NaN         NaN   
168835        0.0                 NaN           NaN         NaN   
168836        3.0                 NaN           NaN         NaN   

        new_deaths_smoothed  ...  female_smokers  male_smokers  \
168832                  NaN  ...            19.1          24.6   
168833                  NaN  ...            19.1          24.6   
168834                  NaN  ...            19.1          24.6   
168835                  NaN  ...            19.1          24.6   
168836                  NaN  ...            19.1          24.6   

        handwashing_facilities  hospital_beds_per_thousand  life_expectancy  \
168832                     NaN                        2.77            78.86   
168833                     NaN                        2.77            78.86   
168834                     NaN                        2.77            78.86   
168835                     NaN                        2.77            78.86   
168836                     NaN                        2.77            78.86   

        human_development_index  excess_mortality_cumulative_absolute  \
168832                    0.926                                   NaN   
168833                    0.926                                   NaN   
168834                    0.926                                   NaN   
168835                    0.926                                   NaN   
168836                    0.926                               -8926.0   

        excess_mortality_cumulative  excess_mortality  \
168832                          NaN               NaN   
168833                          NaN               NaN   
168834                          NaN               NaN   
168835                          NaN               NaN   
168836                        -3.59             -2.63   

        excess_mortality_cumulative_per_million  
168832                                      NaN  
168833                                      NaN  
168834                                      NaN  
168835                                      NaN  
168836                               -26.811643  

[5 rows x 67 columns]

In [9]:
USdf.location.unique()

array(['United States'], dtype=object)

In [10]:
USdf.shape

(815, 67)

The dataset is substantially reduced in size as it now has 815 records. This is still acceptable as this should be enough samples to give a statistically significant output. Let us now take a look at the statistics again.

In [11]:
USdf.describe()

total_cases     new_cases  new_cases_smoothed   total_deaths  \
count  8.150000e+02  8.140000e+02          809.000000     777.000000   
mean   2.807672e+07  9.903278e+04        99494.343276  472375.917632   
std    2.426983e+07  1.445723e+05       132616.566951  294744.508023   
min    1.000000e+00  0.000000e+00            0.000000       1.000000   
25%    5.234692e+06  2.787325e+04        29550.714000  190786.000000   
50%    2.899251e+07  5.444600e+04        60790.286000  540235.000000   
75%    4.282407e+07  1.170115e+05       120789.286000  699634.000000   
max    8.061268e+07  1.383795e+06       807718.286000  988558.000000   

        new_deaths  new_deaths_smoothed  total_cases_per_million  \
count   777.000000           772.000000               815.000000   
mean   1272.275418          1278.163376             84335.980707   
std     972.272395           775.323753             72900.971153   
min       0.000000             1.714000                 0.003000   
25%     517.000000           716.393250             15723.804500   
50%    1001.000000          1057.000000             87086.805000   
75%    1800.000000          1777.107250            128633.627000   
max    4431.000000          3389.714000            242141.877000   

       new_cases_per_million  new_cases_smoothed_per_million  \
count             814.000000                      809.000000   
mean              297.471591                      298.858031   
std               434.261711                      398.349531   
min                 0.000000                        0.000000   
25%                83.724750                       88.764000   
50%               163.543500                      182.600000   
75%               351.475500                      362.823000   
max              4156.601000                     2426.199000   

       total_deaths_per_million  ...  female_smokers  male_smokers  \
count                777.000000  ...    8.150000e+02  8.150000e+02   
mean                1418.908163  ...    1.910000e+01  2.460000e+01   
std                  885.344441  ...    2.488427e-13  5.332343e-14   
min                    0.003000  ...    1.910000e+01  2.460000e+01   
25%                  573.077000  ...    1.910000e+01  2.460000e+01   
50%                 1622.741000  ...    1.910000e+01  2.460000e+01   
75%                 2101.539000  ...    1.910000e+01  2.460000e+01   
max                 2969.400000  ...    1.910000e+01  2.460000e+01   

       handwashing_facilities  hospital_beds_per_thousand  life_expectancy  \
count                     0.0                8.150000e+02     8.150000e+02   
mean                      NaN                2.770000e+00     7.886000e+01   
std                       NaN                1.555267e-14     4.834658e-13   
min                       NaN                2.770000e+00     7.886000e+01   
25%                       NaN                2.770000e+00     7.886000e+01   
50%                       NaN                2.770000e+00     7.886000e+01   
75%                       NaN                2.770000e+00     7.886000e+01   
max                       NaN                2.770000e+00     7.886000e+01   

       human_development_index  excess_mortality_cumulative_absolute  \
count             8.150000e+02                          1.090000e+02   
mean              9.260000e-01                          4.837607e+05   
std               8.220695e-15                          3.241388e+05   
min               9.260000e-01                         -1.367900e+04   
25%               9.260000e-01                          1.938106e+05   
50%               9.260000e-01                          5.733646e+05   
75%               9.260000e-01                          6.929886e+05   
max               9.260000e-01                          1.065799e+06   

       excess_mortality_cumulative  excess_mortality  \
count                   109.000000        109.000000   
mean                     12.358716         17.366239   
std               

In [12]:
USdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 815 entries, 168832 to 169646
Data columns (total 67 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   iso_code                                    815 non-null    object 
 1   continent                                   815 non-null    object 
 2   location                                    815 non-null    object 
 3   date                                        815 non-null    object 
 4   total_cases                                 815 non-null    float64
 5   new_cases                                   814 non-null    float64
 6   new_cases_smoothed                          809 non-null    float64
 7   total_deaths                                777 non-null    float64
 8   new_deaths                                  777 non-null    float64
 9   new_deaths_smoothed                         772 non-null    float64
 10  total_

The mean of quite a number of variables have changed. This should work in our favor as it is the whole reason why we chose to scope the project to only covid data from the United States. Limiting the scope like this should make the project much less complex.

We also note that there are some columns which do not have any values at all. It is best to get rid of these columns. These would be weekly admissions, weekly admissions per million, and handwashing facilities. Along with these, it would also be best to drop any variables that communicate that the location is in the United states since we have already made sure that the data is only based on the United States.

In [13]:
USdf['iso_code'].nunique()

1

In [14]:
USdf = USdf.drop(['iso_code','continent', 'location'], axis=1)

In [15]:
USdf = USdf.drop(['weekly_icu_admissions', 'weekly_icu_admissions_per_million', 'handwashing_facilities'], axis=1)

In [16]:
USdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 815 entries, 168832 to 169646
Data columns (total 61 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   date                                        815 non-null    object 
 1   total_cases                                 815 non-null    float64
 2   new_cases                                   814 non-null    float64
 3   new_cases_smoothed                          809 non-null    float64
 4   total_deaths                                777 non-null    float64
 5   new_deaths                                  777 non-null    float64
 6   new_deaths_smoothed                         772 non-null    float64
 7   total_cases_per_million                     815 non-null    float64
 8   new_cases_per_million                       814 non-null    float64
 9   new_cases_smoothed_per_million              809 non-null    float64
 10  total_

In [17]:
USdf.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 815 entries, 168832 to 169646
Data columns (total 61 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   date                                        815 non-null    object 
 1   total_cases                                 815 non-null    float64
 2   new_cases                                   814 non-null    float64
 3   new_cases_smoothed                          809 non-null    float64
 4   total_deaths                                777 non-null    float64
 5   new_deaths                                  777 non-null    float64
 6   new_deaths_smoothed                         772 non-null    float64
 7   total_cases_per_million                     815 non-null    float64
 8   new_cases_per_million                       814 non-null    float64
 9   new_cases_smoothed_per_million              809 non-null    float64
 10  total_

Okay, at this stage, the dataset looks pretty good though there is still quite a bit of work to be done. There are 60 features in the dataset and most of them have some null values. The total number of records as seen above is 815. Before we address those missing values, which will be a tedious process, let us look at the features themselves and see how relevant they are to the study.

Remember that we are looking to find the risk factors for Covid Deaths. In other words, we are trying to find the features that are the best predictors for Covid Deaths. This means that number of Covid deaths in a day shall be the target variable. This should be the variable total deaths as can be seen below:

In [22]:
USdf.reset_index(inplace=True)

In [25]:
USdf.drop('index', axis=1, inplace=True)

In [27]:
USdf.drop('level_0', axis=1, inplace=True)

In [28]:
USdf.head()

date  total_cases  new_cases  new_cases_smoothed  total_deaths  \
0  2020/01/22          1.0        NaN                 NaN           NaN   
1  2020/01/23          1.0        0.0                 NaN           NaN   
2  2020/01/24          2.0        1.0                 NaN           NaN   
3  2020/01/25          2.0        0.0                 NaN           NaN   
4  2020/01/26          5.0        3.0                 NaN           NaN   

   new_deaths  new_deaths_smoothed  total_cases_per_million  \
0         NaN                  NaN                    0.003   
1         NaN                  NaN                    0.003   
2         NaN                  NaN                    0.006   
3         NaN                  NaN                    0.006   
4         NaN                  NaN                    0.015   

   new_cases_per_million  new_cases_smoothed_per_million  ...  \
0                    NaN                             NaN  ...   
1                  0.000                             NaN  ...   
2                  0.003                             NaN  ...   
3                  0.000                             NaN  ...   
4                  0.009                             NaN  ...   

   diabetes_prevalence  female_smokers  male_smokers  \
0                10.79            19.1          24.6   
1                10.79            19.1          24.6   
2                10.79            19.1          24.6   
3                10.79            19.1          24.6   
4                10.79            19.1          24.6   

   hospital_beds_per_thousand  life_expectancy  human_development_index  \
0                        2.77            78.86                    0.926   
1                        2.77            78.86                    0.926   
2                        2.77            78.86                    0.926   
3                        2.77            78.86                    0.926   
4                        2.77            78.86                    0.926   

   excess_mortality_cumulative_absolute  excess_mortality_cumulative  \
0                                   NaN                          NaN   
1                                   NaN                          NaN   
2                                   NaN                          NaN   
3                                   NaN                          NaN   
4                               -8926.0                        -3.59   

   excess_mortality  excess_mortality_cumulative_per_million  
0               NaN                                      NaN  
1               NaN                                      NaN  
2               NaN                                      NaN  
3               NaN                                      NaN  
4             -2.63                               -26.811643  

[5 rows x 61 columns]

In [30]:
USdf[USdf['total_deaths'].isnull()]

date  total_cases  new_cases  new_cases_smoothed  total_deaths  \
0   2020/01/22          1.0        NaN                 NaN           NaN   
1   2020/01/23          1.0        0.0                 NaN           NaN   
2   2020/01/24          2.0        1.0                 NaN           NaN   
3   2020/01/25          2.0        0.0                 NaN           NaN   
4   2020/01/26          5.0        3.0                 NaN           NaN   
5   2020/01/27          5.0        0.0                 NaN           NaN   
6   2020/01/28          5.0        0.0               0.571           NaN   
7   2020/01/29          6.0        1.0               0.714           NaN   
8   2020/01/30          6.0        0.0               0.714           NaN   
9   2020/01/31          8.0        2.0               0.857           NaN   
10  2020/02/01          8.0        0.0               0.857           NaN   
11  2020/02/02          8.0        0.0               0.429           NaN   
12  2020/02/03         11.0        3.0               0.857           NaN   
13  2020/02/04         11.0        0.0               0.857           NaN   
14  2020/02/05         11.0        0.0               0.714           NaN   
15  2020/02/06         12.0        1.0               0.857           NaN   
16  2020/02/07         12.0        0.0               0.571           NaN   
17  2020/02/08         12.0        0.0               0.571           NaN   
18  2020/02/09         12.0        0.0               0.571           NaN   
19  2020/02/10         12.0        0.0               0.143           NaN   
20  2020/02/11         13.0        1.0               0.286           NaN   
21  2020/02/12         13.0        0.0               0.286           NaN   
22  2020/02/13         14.0        1.0               0.286           NaN   
23  2020/02/14         14.0        0.0               0.286           NaN   
24  2020/02/15         14.0        0.0               0.286           NaN   
25  2020/02/16         14.0        0.0               0.286           NaN   
26  2020/02/17         14.0        0.0               0.286           NaN   
27  2020/02/18         14.0        0.0               0.143           NaN   
28  2020/02/19         14.0        0.0               0.143           NaN   
29  2020/02/20         14.0        0.0               0.000           NaN   
30  2020/02/21         16.0        2.0               0.286           NaN   
31  2020/02/22         16.0        0.0               0.286           NaN   
32  2020/02/23         16.0        0.0               0.286           NaN   
33  2020/02/24         16.0        0.0               0.286           NaN   
34  2020/02/25         16.0        0.0               0.286           NaN   
35  2020/02/26         16.0        0.0               0.286           NaN   
36  2020/02/27         17.0        1.0               0.429           NaN   
37  2020/02/28         17.0        0.0               0.143           NaN   

    new_deaths  new_deaths_smoothed  total_cases_per_million  \
0          NaN                  NaN                    0.003   
1          NaN                  NaN                    0.003   
2          NaN                  NaN                    0.006   
3          NaN                  NaN                    0.006   
4          NaN                  NaN                    0.015   
5          NaN                  NaN                    0.015   
6          NaN                  NaN                    0.015   
7          NaN                  NaN                    0.018   
8          NaN                  NaN                    0.018   
9          NaN                  NaN                    0.024   
10         NaN                  NaN                    0.024   
11         NaN                  NaN                    0.024   
12         NaN                  NaN                    0.033   
13         NaN                  NaN                    0.033   
14         NaN                  NaN                    0.033   
15         NaN       

As can be seen above, there are only 38 rows missing for total deaths, very likely because there are no recorded deaths up until the 28th of February 2020. total_deaths variable does seem to be a cumulative variable. Let us check to confirm this. We may be looking for a daily count of deaths at this point.

In [31]:
USdf[USdf['total_deaths'].notnull()].head(10)

date  total_cases  new_cases  new_cases_smoothed  total_deaths  \
38  2020/02/29         25.0        8.0               1.286           1.0   
39  2020/03/01         32.0        7.0               2.286           1.0   
40  2020/03/02         55.0       23.0               5.571           6.0   
41  2020/03/03         74.0       19.0               8.286           7.0   
42  2020/03/04        107.0       33.0              13.000          11.0   
43  2020/03/05        184.0       77.0              23.857          12.0   
44  2020/03/06        237.0       53.0              31.429          14.0   
45  2020/03/07        403.0      166.0              54.000          17.0   
46  2020/03/08        519.0      116.0              69.571          21.0   
47  2020/03/09        594.0       75.0              77.000          22.0   

    new_deaths  new_deaths_smoothed  total_cases_per_million  \
38         1.0                  NaN                    0.075   
39         0.0                  NaN                    0.096   
40         5.0                  NaN                    0.165   
41         1.0                  NaN                    0.222   
42         4.0                  NaN                    0.321   
43         1.0                1.714                    0.553   
44         2.0                2.000                    0.712   
45         3.0                2.286                    1.211   
46         4.0                2.857                    1.559   
47         1.0                2.286                    1.784   

    new_cases_per_million  new_cases_smoothed_per_million  ...  \
38                  0.024                           0.004  ...   
39                  0.021                           0.007  ...   
40                  0.069                           0.017  ...   
41                  0.057                           0.025  ...   
42                  0.099                           0.039  ...   
43                  0.231                           0.072  ...   
44                  0.159                           0.094  ...   
45                  0.499                           0.162  ...   
46                  0.348                           0.209  ...   
47                  0.225                           0.231  ...   

    diabetes_prevalence  female_smokers  male_smokers  \
38                10.79            19.1          24.6   
39                10.79            19.1          24.6   
40                10.79            19.1          24.6   
41                10.79            19.1          24.6   
42                10.79            19.1          24.6   
43                10.79            19.1          24.6   
44                10.79            19.1          24.6   
45                10.79            19.1          24.6   
46                10.79            19.1          24.6   
47                10.79            19.1          24.6   

    hospital_beds_per_thousand  life_expectancy  human_development_index  \
38                        2.77            78.86                    0.926   
39                        2.77            78.86                    0.926   
40                        2.77            78.86                    0.926   
41                        2.77            78.86                    0.926   
42                        2.77            78.86                    0.926   
43                        2.77            78.86                    0.926   
44                        2.77            78.86                    0.926   
45                        2.77            78.86                    0.926   
46                        2.77            78.86                    0.926   
47                        2.77            78.86                    0.926   

    excess_mortality_cumulative_absolute  excess_mortality_cumulative  \
38                                   NaN                          NaN   
39                              -13622.6                        -2.48   
40                                   NaN                 

As suspected, the total_deaths column is rather cumulative. Not to necessarily say it wont be useful, but at this point, our more relevant variable will be new_deaths. This will be the variable we look to predict based on a set of independent variables.

This means the other variables related to deaths will not be necessary. This includes the smoothed variables as well. These will now be removed from the dataset.

In [32]:
USdf.drop(['total_deaths', 'new_deaths_smoothed'], axis=1, inplace=True)

In [33]:
USdf.drop(['new_cases_smoothed', 'new_cases_smoothed_per_million', 'new_deaths_smoothed_per_million'], axis=1, inplace=True)

In [34]:
USdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 815 entries, 0 to 814
Data columns (total 56 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   date                                        815 non-null    object 
 1   total_cases                                 815 non-null    float64
 2   new_cases                                   814 non-null    float64
 3   new_deaths                                  777 non-null    float64
 4   total_cases_per_million                     815 non-null    float64
 5   new_cases_per_million                       814 non-null    float64
 6   total_deaths_per_million                    777 non-null    float64
 7   new_deaths_per_million                      777 non-null    float64
 8   reproduction_rate                           765 non-null    float64
 9   icu_patients                                639 non-null    float64
 10  icu_patients_p

In [36]:
USdf.drop(['new_tests_smoothed', 'new_tests_smoothed_per_thousand', 'new_people_vaccinated_smoothed_per_hundred', 'new_people_vaccinated_smoothed', 'new_vaccinations_smoothed_per_million', 'new_vaccinations_smoothed'], axis=1, inplace=True)

In [37]:
USdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 815 entries, 0 to 814
Data columns (total 50 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   date                                     815 non-null    object 
 1   total_cases                              815 non-null    float64
 2   new_cases                                814 non-null    float64
 3   new_deaths                               777 non-null    float64
 4   total_cases_per_million                  815 non-null    float64
 5   new_cases_per_million                    814 non-null    float64
 6   total_deaths_per_million                 777 non-null    float64
 7   new_deaths_per_million                   777 non-null    float64
 8   reproduction_rate                        765 non-null    float64
 9   icu_patients                             639 non-null    float64
 10  icu_patients_per_million                 639 non-n

Since the whole dataset is only on United States, features like gdp_per_capita and smokers can be deleted as they all have the same value. The whole column remains the same and so they cannot act as predictors for the target variable of new_deaths.

In [51]:
USdf['population_density'].unique()

array([35.608])

In [52]:
USdf.drop(['gdp_per_capita', 'female_smokers', 'male_smokers', 'life_expectancy', 'hospital_beds_per_thousand', 'diabetes_prevalence', 'cardiovasc_death_rate', 'extreme_poverty', 'aged_65_older', 'aged_70_older', 'median_age', 'population', 'population_density'], axis=1, inplace=True)

In [54]:
USdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 815 entries, 0 to 814
Data columns (total 37 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   date                                     815 non-null    object 
 1   total_cases                              815 non-null    float64
 2   new_cases                                814 non-null    float64
 3   new_deaths                               777 non-null    float64
 4   total_cases_per_million                  815 non-null    float64
 5   new_cases_per_million                    814 non-null    float64
 6   total_deaths_per_million                 777 non-null    float64
 7   new_deaths_per_million                   777 non-null    float64
 8   reproduction_rate                        765 non-null    float64
 9   icu_patients                             639 non-null    float64
 10  icu_patients_per_million                 639 non-n

In [59]:
for i in USdf.columns:
    if len(USdf[i].unique()) == 1:
        print(i)

human_development_index


In [60]:
USdf.drop(['human_development_index'], axis=1, inplace=True)

In [65]:
USdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 815 entries, 0 to 814
Data columns (total 36 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   date                                     815 non-null    object 
 1   total_cases                              815 non-null    float64
 2   new_cases                                814 non-null    float64
 3   new_deaths                               777 non-null    float64
 4   total_cases_per_million                  815 non-null    float64
 5   new_cases_per_million                    814 non-null    float64
 6   total_deaths_per_million                 777 non-null    float64
 7   new_deaths_per_million                   777 non-null    float64
 8   reproduction_rate                        765 non-null    float64
 9   icu_patients                             639 non-null    float64
 10  icu_patients_per_million                 639 non-n

At this point we have trimmed the dataset down quite a bit. But there are still a lot of variables with null values. It is now time to take a closer look at these and see if we can deal with these missing values.

In [66]:
USdf[['date', 'new_cases']].head()

date  new_cases
0  2020/01/22        NaN
1  2020/01/23        0.0
2  2020/01/24        1.0
3  2020/01/25        0.0
4  2020/01/26        3.0

In [67]:
USdf[['date', 'new_deaths']]

date  new_deaths
0    2020/01/22         NaN
1    2020/01/23         NaN
2    2020/01/24         NaN
3    2020/01/25         NaN
4    2020/01/26         NaN
..          ...         ...
810  2022/04/11       322.0
811  2022/04/12       561.0
812  2022/04/13       956.0
813  2022/04/14       778.0
814  2022/04/15       437.0

[815 rows x 2 columns]

Total Cases and Total Deaths were only missing in the first few rows. It is reasonable to assume that for those day's entries, there had not yet been any cases or deaths, so the missing values can be filled with zeros.The same applies for the same stats per million.

In [75]:
USdf.fillna({'new_cases':0, 'new_deaths':0, 'total_cases_per_million':0, 'new_cases_per_million':0, 'total_deaths_per_million':0, 'new_deaths_per_million':0})

In [76]:
USdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 815 entries, 0 to 814
Data columns (total 36 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   date                                     815 non-null    object 
 1   total_cases                              815 non-null    float64
 2   new_cases                                815 non-null    float64
 3   new_deaths                               815 non-null    float64
 4   total_cases_per_million                  815 non-null    float64
 5   new_cases_per_million                    815 non-null    float64
 6   total_deaths_per_million                 815 non-null    float64
 7   new_deaths_per_million                   815 non-null    float64
 8   reproduction_rate                        765 non-null    float64
 9   icu_patients                             639 non-null    float64
 10  icu_patients_per_million                 639 non-n

The Covid 19 reproduction rate informs us about the spread of Covid 19 as it is passed on from one person to another. These will also be filled with zeros

In [95]:
USdf.fillna({'icu_patients':0, 'icu_patients_per_million':0, 'hosp_patients':0, 'hosp_patients_per_million':0, 'weekly_hosp_admissions':0, 'weekly_hosp_admissions_per_million':0}, inplace=True)

In [96]:
USdf.fillna({'total_tests':0, 'total_tests_per_thousand':0, 'new_tests':0, 'new_tests_per_thousand':0, 'tests_per_case':0, 'positive_rate':0}, inplace=True)

In [97]:
USdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 815 entries, 0 to 814
Data columns (total 36 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   date                                     815 non-null    object 
 1   total_cases                              815 non-null    float64
 2   new_cases                                815 non-null    float64
 3   new_deaths                               815 non-null    float64
 4   total_cases_per_million                  815 non-null    float64
 5   new_cases_per_million                    815 non-null    float64
 6   total_deaths_per_million                 815 non-null    float64
 7   new_deaths_per_million                   815 non-null    float64
 8   reproduction_rate                        815 non-null    float64
 9   icu_patients                             815 non-null    float64
 10  icu_patients_per_million                 815 non-n

In [98]:
USdf.fillna({'tests_per_case':0, 'total_vaccinations':0, 'people_vaccinated':0, 'people_fully_vaccinated':0, 'new_vaccinations':0, 'total_vaccinations_per_hundred':0, 'people_vaccinated_per_hundred':0, 'people_fully_vaccinated_per_hundred':0, 'total_boosters_per_hundred':0}, inplace=True)

In [99]:
USdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 815 entries, 0 to 814
Data columns (total 36 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   date                                     815 non-null    object 
 1   total_cases                              815 non-null    float64
 2   new_cases                                815 non-null    float64
 3   new_deaths                               815 non-null    float64
 4   total_cases_per_million                  815 non-null    float64
 5   new_cases_per_million                    815 non-null    float64
 6   total_deaths_per_million                 815 non-null    float64
 7   new_deaths_per_million                   815 non-null    float64
 8   reproduction_rate                        815 non-null    float64
 9   icu_patients                             815 non-null    float64
 10  icu_patients_per_million                 815 non-n

In [106]:
USdf[['date','stringency_index']].isnull().sum()

date                0
stringency_index    8
dtype: int64

In [110]:
USdf['stringency_index'].tail(20)

795    58.8
796    58.8
797    58.8
798    58.8
799    58.8
800    58.8
801    58.8
802    58.8
803    58.8
804    58.8
805    58.8
806    58.8
807     NaN
808     NaN
809     NaN
810     NaN
811     NaN
812     NaN
813     NaN
814     NaN
Name: stringency_index, dtype: float64

As can be seen above, only the last 8 entries don't have a stringency index. Looking at the trend in the data, we will provide a stringency index of 58.8 to these as well. The stringency index is just an indicator of how restrictive the covid policies were eg. lockdowns, mask wearing, etc..

In [111]:
USdf['stringency_index'].fillna(58.8, inplace=True)

In [112]:
USdf['stringency_index'].tail(20)

795    58.8
796    58.8
797    58.8
798    58.8
799    58.8
800    58.8
801    58.8
802    58.8
803    58.8
804    58.8
805    58.8
806    58.8
807    58.8
808    58.8
809    58.8
810    58.8
811    58.8
812    58.8
813    58.8
814    58.8
Name: stringency_index, dtype: float64

Checking whether people_vaccinated column is identical with people_fully_vaccinated column. Since they are not identical, both columns are retained.

In [113]:
USdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 815 entries, 0 to 814
Data columns (total 36 columns):
 #   Column                                   Non-Null Count  Dtype  
---  ------                                   --------------  -----  
 0   date                                     815 non-null    object 
 1   total_cases                              815 non-null    float64
 2   new_cases                                815 non-null    float64
 3   new_deaths                               815 non-null    float64
 4   total_cases_per_million                  815 non-null    float64
 5   new_cases_per_million                    815 non-null    float64
 6   total_deaths_per_million                 815 non-null    float64
 7   new_deaths_per_million                   815 non-null    float64
 8   reproduction_rate                        815 non-null    float64
 9   icu_patients                             815 non-null    float64
 10  icu_patients_per_million                 815 non-n

Excess mortality seems to be missing a whole lot of values. Its also a statistic that involves the current death rate and so it would give a false sense of the predictability/accuracy of the resulting model. Based on this logic, the features for excess mortality will be dropped.

In [116]:
USdf.drop(['excess_mortality', 'excess_mortality_cumulative_absolute', 'excess_mortality_cumulative', 'excess_mortality_cumulative_per_million'], axis=1, inplace=True)

In [117]:
USdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 815 entries, 0 to 814
Data columns (total 32 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   date                                 815 non-null    object 
 1   total_cases                          815 non-null    float64
 2   new_cases                            815 non-null    float64
 3   new_deaths                           815 non-null    float64
 4   total_cases_per_million              815 non-null    float64
 5   new_cases_per_million                815 non-null    float64
 6   total_deaths_per_million             815 non-null    float64
 7   new_deaths_per_million               815 non-null    float64
 8   reproduction_rate                    815 non-null    float64
 9   icu_patients                         815 non-null    float64
 10  icu_patients_per_million             815 non-null    float64
 11  hosp_patients                   

In [127]:
USdf[['date','total_boosters']]

date  total_boosters
0    2020/01/22             NaN
1    2020/01/23             NaN
2    2020/01/24             NaN
3    2020/01/25             NaN
4    2020/01/26             NaN
..          ...             ...
810  2022/04/11      99070369.0
811  2022/04/12      99138381.0
812  2022/04/13      99148999.0
813  2022/04/14             NaN
814  2022/04/15             NaN

[815 rows x 2 columns]

For the total Boosters missing values, we will make the initial missing values zero and then after the first entry, we will make fill with previous entries.

In [130]:
USdf['total_boosters'].fillna(method='ffill', inplace=True)

In [132]:
USdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 815 entries, 0 to 814
Data columns (total 32 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   date                                 815 non-null    object 
 1   total_cases                          815 non-null    float64
 2   new_cases                            815 non-null    float64
 3   new_deaths                           815 non-null    float64
 4   total_cases_per_million              815 non-null    float64
 5   new_cases_per_million                815 non-null    float64
 6   total_deaths_per_million             815 non-null    float64
 7   new_deaths_per_million               815 non-null    float64
 8   reproduction_rate                    815 non-null    float64
 9   icu_patients                         815 non-null    float64
 10  icu_patients_per_million             815 non-null    float64
 11  hosp_patients                   

In [134]:
USdf['tests_units'].unique()

array([nan, 'tests performed'], dtype=object)

In [135]:
USdf['tests_units'].tail(50)

0                 NaN
1                 NaN
2                 NaN
3                 NaN
4                 NaN
5                 NaN
6                 NaN
7                 NaN
8                 NaN
9                 NaN
10                NaN
11                NaN
12                NaN
13                NaN
14                NaN
15                NaN
16                NaN
17                NaN
18                NaN
19                NaN
20                NaN
21                NaN
22                NaN
23                NaN
24                NaN
25                NaN
26                NaN
27                NaN
28                NaN
29                NaN
30                NaN
31                NaN
32                NaN
33                NaN
34                NaN
35                NaN
36                NaN
37                NaN
38                NaN
39    tests performed
40    tests performed
41    tests performed
42    tests performed
43    tests performed
44    tests performed
45    test

In [139]:
USdf.tests_units.tail(50)

765    tests performed
766    tests performed
767    tests performed
768    tests performed
769    tests performed
770    tests performed
771    tests performed
772    tests performed
773    tests performed
774    tests performed
775    tests performed
776    tests performed
777    tests performed
778    tests performed
779    tests performed
780    tests performed
781    tests performed
782    tests performed
783    tests performed
784    tests performed
785    tests performed
786    tests performed
787    tests performed
788    tests performed
789    tests performed
790    tests performed
791    tests performed
792    tests performed
793    tests performed
794    tests performed
795    tests performed
796    tests performed
797    tests performed
798    tests performed
799    tests performed
800    tests performed
801    tests performed
802    tests performed
803    tests performed
804    tests performed
805    tests performed
806    tests performed
807    tests performed
808    test

For tests_units column, it seems its just the first records, and the last records that have missing values. We shall perform foward fill and then fill with zeros

In [140]:
USdf['tests_units'].fillna(method='ffill', inplace=True)

In [141]:
USdf['tests_units'].fillna(0, inplace=True)

In [142]:
USdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 815 entries, 0 to 814
Data columns (total 32 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   date                                 815 non-null    object 
 1   total_cases                          815 non-null    float64
 2   new_cases                            815 non-null    float64
 3   new_deaths                           815 non-null    float64
 4   total_cases_per_million              815 non-null    float64
 5   new_cases_per_million                815 non-null    float64
 6   total_deaths_per_million             815 non-null    float64
 7   new_deaths_per_million               815 non-null    float64
 8   reproduction_rate                    815 non-null    float64
 9   icu_patients                         815 non-null    float64
 10  icu_patients_per_million             815 non-null    float64
 11  hosp_patients                   

In [147]:
USdf[['date','total_boosters']].tail(50)

date  total_boosters
765  2022/02/25      95235120.0
766  2022/02/26      95341799.0
767  2022/02/27      95395465.0
768  2022/02/28      95539878.0
769  2022/03/01      95660131.0
770  2022/03/02      95768586.0
771  2022/03/03      95877724.0
772  2022/03/04      96014896.0
773  2022/03/05      96089745.0
774  2022/03/06      96124939.0
775  2022/03/07      96215375.0
776  2022/03/08      96307317.0
777  2022/03/09      96396353.0
778  2022/03/10      96486628.0
779  2022/03/11      96596658.0
780  2022/03/12      96653898.0
781  2022/03/13      96683256.0
782  2022/03/14      96766181.0
783  2022/03/15      96853074.0
784  2022/03/16      96941521.0
785  2022/03/17      97027890.0
786  2022/03/18      97139705.0
787  2022/03/19      97200964.0
788  2022/03/20      97231769.0
789  2022/03/21      97310028.0
790  2022/03/22      97387161.0
791  2022/03/23      97467328.0
792  2022/03/24      97546018.0
793  2022/03/25      97644917.0
794  2022/03/26      97697264.0
795  2022/03/27      97723922.0
796  2022/03/28      97793446.0
797  2022/03/29      97870016.0
798  2022/03/30      97974874.0
799  2022/03/31      98095277.0
800  2022/04/01      98231898.0
801  2022/04/02      98302064.0
802  2022/04/03      98342109.0
803  2022/04/04      98445774.0
804  2022/04/05      98547148.0
805  2022/04/06      98652592.0
806  2022/04/07      98757821.0
807  2022/04/08      98880064.0
808  2022/04/09      98941040.0
809  2022/04/10      98977001.0
810  2022/04/11      99070369.0
811  2022/04/12      99138381.0
812  2022/04/13      99148999.0
813  2022/04/14      99148999.0
814  2022/04/15      99148999.0

The total boosters feature will also be forward filled then zeros added at the start.

In [148]:
USdf['total_boosters'].fillna(method='ffill', inplace=True)

In [149]:
USdf['total_boosters'].fillna(0, inplace=True)

In [150]:
USdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 815 entries, 0 to 814
Data columns (total 32 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   date                                 815 non-null    object 
 1   total_cases                          815 non-null    float64
 2   new_cases                            815 non-null    float64
 3   new_deaths                           815 non-null    float64
 4   total_cases_per_million              815 non-null    float64
 5   new_cases_per_million                815 non-null    float64
 6   total_deaths_per_million             815 non-null    float64
 7   new_deaths_per_million               815 non-null    float64
 8   reproduction_rate                    815 non-null    float64
 9   icu_patients                         815 non-null    float64
 10  icu_patients_per_million             815 non-null    float64
 11  hosp_patients                   

In [152]:
USdf.duplicated().sum()

0

In [153]:
USdf.duplicated(subset=['people_vaccinated', 'people_fully_vaccinated'])

0      False
1       True
2       True
3       True
4       True
       ...  
810    False
811    False
812    False
813     True
814     True
Length: 815, dtype: bool

In [154]:
vacdup = USdf.duplicated(subset=['people_vaccinated', 'people_fully_vaccinated'])

In [155]:
vacdup[vacdup != True]

0      False
326    False
327    False
328    False
329    False
       ...  
808    False
809    False
810    False
811    False
812    False
Length: 488, dtype: bool

In [156]:
USdf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 815 entries, 0 to 814
Data columns (total 32 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   date                                 815 non-null    object 
 1   total_cases                          815 non-null    float64
 2   new_cases                            815 non-null    float64
 3   new_deaths                           815 non-null    float64
 4   total_cases_per_million              815 non-null    float64
 5   new_cases_per_million                815 non-null    float64
 6   total_deaths_per_million             815 non-null    float64
 7   new_deaths_per_million               815 non-null    float64
 8   reproduction_rate                    815 non-null    float64
 9   icu_patients                         815 non-null    float64
 10  icu_patients_per_million             815 non-null    float64
 11  hosp_patients                   

In [158]:
USdf.to_csv('Covid_Clean_Dataset')